apps/sgx/enclave.cc

```diff
@@ -1,62 +1,32 @@
 #include <dlpack/dlpack.h>
+#include <tvm/runtime/c_runtime_api.h>
 #include <tvm/runtime/module.h>
 #include <tvm/runtime/registry.h>
 #include <tvm/runtime/packed_func.h>
-#ifndef _LIBCPP_SGX_CONFIG
-#include <iostream>
-#endif
-
-/* This function mirrors the one in howto_deploy except without the iostream */
-int Verify(tvm::runtime::Module mod, std::string fname) {
-  // Get the function from the module.
-  tvm::runtime::PackedFunc f = mod.GetFunction(fname);
-
-  // Allocate the DLPack data structures.
-  DLTensor* x;
-  DLTensor* y;
-  int ndim = 1;
-  int dtype_code = kDLFloat;
-  int dtype_bits = 32;
-  int dtype_lanes = 1;
-  int device_type = kDLCPU;
-  int device_id = 0;
-  int64_t shape[1] = {10};
-  TVMArrayAlloc(shape, ndim, dtype_code, dtype_bits, dtype_lanes,
-                device_type, device_id, &x);
-  TVMArrayAlloc(shape, ndim, dtype_code, dtype_bits, dtype_lanes,
-                device_type, device_id, &y);
-  for (int i = 0; i < shape[0]; ++i) {
-    static_cast<float*>(x->data)[i] = i;
-  }
-
-  // Invoke the function
-  f(x, y);
-
-  // check the output
-  bool all_eq = true;
-  for (int i = 0; i < shape[0]; ++i) {
-    all_eq = all_eq && static_cast<float*>(y->data)[i] == i + 1.0f;
-  }
-
-  return all_eq;
-}
+#include "../../src/runtime/sgx/trusted/runtime.h"
+#include "../../src/runtime/sgx/trusted/runtime.cc"
 
+using namespace tvm::runtime;
 
 extern "C" {
-void tvm_ecall_run_module(const void* tvm_args, void* tvm_return_value) {
-  tvm::runtime::Module mod_syslib = (*tvm::runtime::Registry::Get("module._GetSystemLib"))();
-  *(int*)tvm_return_value = Verify(mod_syslib, "addonesys");
-}
+void tvm_ecall_init() {}
+
+void tvm_ecall_packed_func(const char* cname,
+                           void* tvm_args,
+                           void* tvm_ret_val) {
+  std::string name = std::string(cname);
+  CHECK(name.substr(0, sgx::ECALL_PACKED_PFX.size()) == sgx::ECALL_PACKED_PFX)
+    << "Function `" << name << "` is not an enclave export.";
+  const PackedFunc* f = Registry::Get(name);
+  CHECK(f != nullptr) << "Enclave function not found.";
+  f->CallPacked(*reinterpret_cast<TVMArgs*>(tvm_args),
+      reinterpret_cast<TVMRetValue*>(tvm_ret_val));
```

Comment:
This likely won't work, because of ABI. The TVMRetValue is not POD type, if we return something like std::string. By default std::string uses internal allocator in enclave, which the outside is not able to see. 

Use the C level ocall_set_return to provide correct visible memory in the untrusted space in order to make it really work